src/app/components/edit-tool-group-resource/tool-group-resource.component.ts

```diff
@@ -0,0 +1,139 @@
+  @Input() toolGroup: ToolGroup;
+  @Input() resources: Resource[] = [];
+  selectedResources: Resource[] = [];
+  initialTools: Tools[];
+  tools: Tools[];
+  saving = false;
+  errorMessage: string[] = [];
+  ruleData: any;
+
+  constructor(
+    protected toolGroupService: ToolGroupService,
+    protected activeModal: NgbActiveModal,
+    protected resourceService: ResourceService,
+  ) {}
+
+  ngOnInit(): void {
+    this.initialTools = this.toolGroup.tools;
+    if (this.initialTools.length) {
+      this.tools = this.initialTools;
+    } else {
+      this.tools = [this.generateNewResource()];
+    }
+
+    if (!this.resources.length) {
+      this.resourceService
+        .getResources()
+        .then((resources) => {
+          this.resources = resources;
+        })
+        .catch(this.handleError.bind(this));
+    }
+  }
+
+  async createOrUpdate(): Promise<void> {
+    this.saving = true;
+    const promises = [];
+    if (this.tools.length) {
+      this.tools.forEach((tool) => {
+        const isUpdate = !!this.initialTools.find(
+          (item) => item.id === tool.id,
+        );
+        promises.push(
+          this.toolGroupService.addOrUpdateTool(
+            this.toolGroup.id,
+            tool.id,
+            tool.tool.id.toString(),
+            tool.suggestionsWeight,
+            isUpdate,
+          ),
+        );
+      });
+    }
+
+    const results: PromisePayload[] = await Promise.all(
+      promises.map((p) =>
+        p
+          .then((value) => ({
+            success: true,
+            value,
+          }))
+          .catch((error) => ({
+            success: false,
+            error,
+          })),
+      ),
+    );
+
+    const invalidResults = results.filter((result) => !result.success);
+
+    if (invalidResults.length) {
+      this.saving = false;
+      invalidResults.forEach((invalidResult) => {
+        this.errorMessage = [...this.errorMessage, invalidResult.error];
+      });
+    } else {
+      this.activeModal.close('closed');
+    }
+  }
+
+  updateTool(event: Tools): void {
+    const toolIndex = this.tools.findIndex((tool) => tool.id === event.id);
+    const tools = this.tools;
+    tools[toolIndex] = event;
+    this.tools = tools;
+  }
+
+  deleteTool(event): void {
+    this.tools = this.tools.filter((tool) => tool.id !== event);
+    // If ID requested to be deleted an existing tool
+    const shouldDelete = !!this.initialTools.find((item) => item.id === event);
+    if (shouldDelete) {
+      this.toolGroupService.deleteTool(this.toolGroup.id, event);
+    }
+  }
+
+  addResource(): void {
+    this.tools = [...this.tools, this.generateNewResource()];
+  }
+
+  generateNewResource(): Tools {
+    return {
+      id: this.generateId(),
+      suggestionsWeight: '',
+      tool: undefined,
+      'tool-group': this.toolGroup,
+    };
+  }
+
+  generateId(): string {
+    return Math.floor(Math.random() * 1000).toString();
```

Comment:
In the past when I've needed to generate a new temporary id and avoid potential conflicts with ids returned from the API, I've used a random negative number. This works because the ids the database generates are all positive.

I'm not sure if your random 0-999 id generated here will have a conflict or not, but using negatives could add an extra layer of protection against collisions

Reply:
also, within a jsonapi data model universe, the `tuple(type, id)` should be unique, and if 2 objects share that tuple they should be considered the same object.

This has implications with the underlying jsonapi parser/cache that we are using for mobile-content-admin. Right now we are only really using the parser aspect and not really relying on the cache aspect. Long term it might be beneficial to utilize the in memory cache as we improve admin performance

Reply:
I'm going to utilise `angular2-uuid` to create a UUID, to avoid potential conflicts.

You're right, there are many ways, we could take advantage of `JsonApiDataStore()` and using the `. find()` function to read the saved data.